### DISCUSSION TITLE
Performance Testing

### DISCUSSION THREAD

**Discussion by zackarno** [OCHA-DAP/pa-aa-yem-flooding#7] [General]
Starting this discussion to start a discussion on performance testing for the a potential Yemen flooding trigger. Would be great to get opinions on best metrics as well as idea on how best to classify TP, TN, FP, FN.

I suppose minimizing FNs would be the most important consideration?

TNs seem to be the most difficult to confidently classify. I've illustrated some of the issues with these classifications in the drawing below.

![image](https://user-images.githubusercontent.com/23731112/223358846-92db03bb-4f15-42ef-a407-4f48ed2f2ca8.png)

Currently we don't have a good method for classifying the TNs, so are thinking to just classify TP, FP, FNs by looking at both reported flood dates and dates/sites where the threshold was crossed.

- if reported flood event is below threshold -> FN
- if threshold crossed and no flood event -> FP
- if threshold crossed and flood event -> TP

Even with this simplified classification there is is still some decisions of what to do after a TP if the precip keeps increasing or decreases and then increases. So far we are thinking to remove any consecutive values after a TP that are at or above the threshold. However, we should probably impose a sort of maximum consecutive day limitation on this and figure out what to do if it decreases for a short period and then continues to decrease

**Comment by caldwellst** [comment]
Nice images!! How do you draw and write so cleanly? Question, is the data above mapping flood extents? I'm assuming so below.

### True negatives

The conflict literature's approach to onset prediction essentially has to deal with this same problem, and is fresh on my mind! In general, I think your approach makes sense. TN would be oversampled anyway, and would drive performance measures like specificity and accuracy, so better to rely precision and recall (and AUPRC) instead of using accuracy or specificity (or AUROC). I also believe that our trigger report cards don't include any measures relying on TN anyway?

### Events

For conflict, the events that define the other three parts of the confusion matrix are determined by what onset is (obviously). They define onset as conflict reaching a certain threshold after a period of time where the threshold had not been met. So, onset might be the first month with 25 conflict-related fatalities after 6 months with fewer than 25 conflict-related fatalities. Months of conflict after the onset are considered NA, as are months not reaching the conflict threshold if they are not in an extended period of non-conflict (the 6 months above).

So in the case of floods, I was wondering if you might consider singular events in the same way. Since it's unlikely that the reported flood events are just on the singular dates, consider a single flood to be one that lasts until levels reset. This is assuming you're matching up with Floodscan data, but the same logic could similarly apply to the forecasts, that in the second plot you're merely capturing a single flooding event (that does grow and change). 

### Dropping data

Now, if you define the flood as an extended event, it raises the question of how to drop data. You could drop anything after the first date, but then are we removing the possibility of detecting it before the event is over/peaked, but not necessarily by the first date? You could consider if you would trigger any time before the peak or halfway point within a single event (if you define it as above)? That would be a TP, and you could have statistics on how early or late into an event they have been detected.

**Reply by zackarno** [reply]
good points - responding below. I wish I could draw by hand that neatly -- I used [excalidraw](https://excalidraw.com/). It's just a website which is awesome and easy for this kind of stuff + you can collab on sketches

**Reply by PaulineNimo** [reply]
Good points Seth. TN can be tricky. Would every time point without an event be classified as a TN? That greatly distorts the performance metrics. 

For floods, would the threshold be said to be the time point during which the water levels were high enough to cause damage? At that point, the damage is done and any further increase in water levels would only compound the effect?

**Reply by zackarno** [reply]
yeah i think TN is too tricky that we will just not include it in the performance testing yet (or maybe not at all). Yes good point : if the threshold is only TP when a flood event is reported it's likely that at this point the damage is done.  But eventually we will be switching to rain forecasting in which case we will have lead time to prepare for an event that would cause the flooding. Nonetheless a system for earlier detection would be beneficial:  in the graph below the vertical dashed line should  probably represent a TP in this case.

**Comment by zackarno** [comment]
Another situation: 

<img width="652" alt="image" src="https://user-images.githubusercontent.com/23731112/223423077-6fbfceda-91c2-4621-8103-37ef6aaddd3a.png">

**Reply by caldwellst** [reply]
Yeah, this is why I think you might want to extend your temporal definition of when an event is and then measure how you can detect it. In this case above, maybe it's TP up to the peak and FP beyond. You could produce additional metrics on TP based on how early floods are being detect + on FPs, essentially if floods are detected but too late (after the peak maybe?) or not detected at all.

**Reply by PaulineNimo** [reply]
Would it not be a TP until it crosses the threshold after the peak (on its way down)?

**Reply by zackarno** [reply]
@PaulineNimo  - yes I think so, but we just want to count it 1x as 1 TP. Then I think we still don't think we want to count dates after peak but above threshold as FP, so  in the example above, I think it would be 1 TP and no FP or FN

**Comment by zackarno** [comment]
Your points above make sense - for simplicity, let's leave flood scan out of it for the moment and just think about historical rainfall to try to establish a cause-effect between how much it rained and whether or not flooding was reported.

The hard parts seem to be a.) determining when flood levels reset, b.) how far to look back or forward. Here is another plot with some real data for one site:

I've drawn two thresholds (in blue), but as you know, we have to calculate these performance metrics for every threshold

![image](https://user-images.githubusercontent.com/23731112/223649352-ed4afcb3-e069-4c99-883e-8dd3bf206a72.png)

so let's look at the 25 mm threshold:

- We could call Aug 9 a TP, but what about date July 15- Aug 8? Also what about Aug 10 onwards?
    + I think we agree that we can remove Aug 10 -> Sep 02 from consideration as FPs
    + I could see a case for July 16,17, 19 being FPs, but July 27- Aug 8 look to me more like they should not be FPs
- it's tricky because every situation/site will be very different so I am trying to think of how we could scientifically come up with less subjective rules
    + would it be helpful if at every threshold crossing I calculated the number of days forward to the next flood event as well backward to the previous flood event. I wonder if this could somehow be used?

**Reply by PaulineNimo** [reply]
How about looking at how many time points for each crossing of the threshold before dipping again? Also, to your second point, do we have an idea of roughly how long each flood event is?

**Reply by zackarno** [reply]
@PaulineNimo  - I like that idea of counting how many days for each crossing of threshold.

We don't have an idea of how long the flood event are from the flood db unfortunately

**Reply by caldwellst** [reply]
Yeah, exactly Pauline, agree on needing a definition of how long an event is. Then I would define how early we want to respond to it, which might require knowledge of what the actual response would be on the ground, and how early is too early (becoming a false alarm).

For triggering after the event, how many events are occurring multiple times in a single season in the same location? How far apart are these instances? What does the pattern look like there? Looking at the slide deck, it appears to me that Yemen has a unimodal rainy season, and that the signals to respond to these flood events aren't easily separable (those within the peaks fall generally around the exact same peaks).

Would it make more sense to transform the problem from a granular temporal event to triggering whether a specific site will experience flooding in that year? Looking at those massive peaks, almost every site with these massive peaks saw flooding events. This might provide an easier framework to improve performance because it doesn't look from these plots that timing the flooding based on precipitation is particularly easy, but knowing that one might occur is a much simpler problem. As well, this might fit better into the operational realities of Yemen where early action is still potentially going to take a while?

**Reply by zackarno** [reply]
true to both:
1. a definition of how long events are would be useful 
2. There are cases of multiple reported events that do seem to occur within short-spans of time that make the rainfall event hard to separate. 

On pt 2 however, this might be less of an issue than it looks in the slide decks which is often showing just top 10 sites (by # shelters damaged). If you look at the all of data the vast majority of sites report 1 event per year

<img width="612" alt="image" src="https://user-images.githubusercontent.com/23731112/223671411-a2668be1-4ff5-413a-b366-c2d1c13b7217.png">


transforming the problem to whether or not these sites are more likely to experience flooding does certainly simplify the matter. If that's all we need then we just say that all the sites classified as "High Flood Risk" are more likely, done. This could be enough, but then if we get below average rainfall it could be a big false alarm? I'm not sure we can say whether or not it flooding is likely without a some estimate/prediction of rainfall quantity.

The flooding shown in the visuals/slide deck show flooding linked to abnormally high rainfall, yes there is a seasonal aspect, but these events are anomalous.

**Reply by caldwellst** [reply]
I wasn't talking about site risk, but a classification problem across a wider temporal scale, so you predict exactly if a site will experience a flood in the next x period of time, and calculate TP, FP, etc. accordingly.

**Reply by caldwellst** [reply]
Essentially, monitoring for a threshold that if met any time in the season (could be a varying threshold based on time in the season even), it activates for that site, but never can activate again that season. This doesn't preclude looking outside the rainy season, where you could also bin months together into multiple non-rainy seasons or monitoring per month or whatever (maybe signals are much clearer when rain is sparser).

**Reply by zackarno** [reply]
got it! that could be a good approach

**Comment by zackarno** [comment]
synthesizing a few points from slack:

- we currently only have sites that reported flooding not sites that didn't. Problem bc it's possible there are sites that didn't flood with the same rainfall patterns. 
    + a comprehensive list of sites where the flood reporting/monitoring system was rolled out would be beneficial (i've emailed to find out if possible)
    + without knowing where flooding did NOT occur performance testing is more difficult and it might be site-level conditions that are actually driving flood more than rainfall.
- This being said I think rainfall is still a primary factor behind the quantity of water in the system -- maybe site-level factors affect flooding such as drainage and topography... but still in the sites that DID report flooding : x mm of rainfall + conditions (which we assume don't change like topography) led to the flooding and we could potentially design the trigger/thresholds around that?

Below is another plot illustrating a. ) trickiness of classification, b.) potentially seths point regarding the potential for factors other than rainfall to drive flood occurence. It has a similar rainfall characteristics as above, but has **2** flood events rather than 1, why!?

<img width="1164" alt="image" src="https://user-images.githubusercontent.com/23731112/223723931-29684c8a-9e1f-49f0-926a-58dd07ddc128.png">

**Comment by turnerm** [comment]
Lots of good discussions here, just jumping in with info on what we've done for previous pilots (Bangladesh and Nepal), which I think aligns with what everyone has been saying.

The main point is that we want to understand the performance from the perspective of a functional trigger. If we follow our previous approach, then the rainfall timeseries (which was a river discharge timeseries for the riverine pilot analogues) should be transformed into a list of dates, by taking the dates when the threshold is crossed, which represent the occurrence of a trigger. Then, these dates can be combined with the list of flood event dates to get the performance metrics. For the previous pilots we used a window around the flood event date that was somewhat generous in terms of early activation and cut off at the event date.
